### PR TITLE
ARROW-13642: [C++][Compute] Hash join node supporting all semi, anti, inner, outer join types

### PR DESCRIPTION
**Pull request by michalursa**
Hash join node implementation.

Complete implementation of in-memory hash join, except for:
- residual predicates (non-equality join conditions accompanying equality conditions in case of outer joins)
- support for dictionary data types.

This is a simple implementation that is not performance optimized. More advanced implementation (adding e.g. Bloom-like filters) will follow.

Supports all 8 flavors of the join: left semi, left anti-semi, right semi, right anti-semi, inner, left outer, right outer and full outer join.
Note that left semi join, for instance, is not the same as right semi join with inputs swapped. They have the same effect but the algorithm is different in both cases. 
Left semi join will filter out rows with no matches coming from left (first) child exec node. 
Right semi join will keep track of whether there was a match for each row for right (second) child exec node inserted into a hash table and output hash table rows with matches at the end.

Supports the same keys as hash group by node, except for lack of support for dictionaries and strings with 64-bit offset.

Supports choosing for each corresponding pair of key fields whether null should match null or not.

### CODE REVIEW

**Review by bkietz — CHANGES_REQUESTED** [review]
just a few comments so far

File: cpp/src/arrow/compute/exec/schema_util.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        }
+      }
+      mapping_bufs_[pos][i] = field_id;
+    }
+    mapping_ptrs_[pos] = mapping_bufs_[pos].data();
+  }
+
+  std::vector<int*> mapping_ptrs_;
+  std::vector<std::vector<int>> mapping_bufs_;
+  std::vector<std::pair<SchemaHandleType, std::vector<FieldInfo>>> schemas_;
```
[the remark this answers is not in the document]
Comment (michalursa) [reply cpp/src/arrow/compute/exec/schema_util.h]: comment added

File: cpp/src/arrow/compute/exec/schema_util.h [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+namespace compute {
+
+// Helper class for managing related row schemas.
+// Used to efficiently map any column in one schema to a corresponding column in another
+// schema if such exists.
+// Materialized mappings are generated lazily at the time of the first access.
+// Thread-safe apart from initialization.
+//
+template <typename SchemaHandleType>
```
[the remark this answers is not in the document]
Comment (michalursa) [reply cpp/src/arrow/compute/exec/schema_util.h]: Changed to ProjectionIdEnum

File: cpp/src/arrow/compute/exec/hash_join_node.cc [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+  void InputReceived(ExecNode* input, ExecBatch batch) override {
+    ARROW_DCHECK(std::find(inputs_.begin(), inputs_.end(), input) != inputs_.end());
+
+    if (finished_.is_finished()) {
+      return;
+    }
+
+    size_t thread_index = thread_indexer_();
+    int side = (input == inputs_[0]) ? 0 : 1;
+    ErrorIfNotOk(impl_->InputReceived(thread_index, side, std::move(batch)));
```
[the remark this answers is not in the document]
Comment (michalursa) [reply cpp/src/arrow/compute/exec/hash_join_node.cc]: I added returns after ErrorIfNotOk calls.

**Pull request closed without merging**